<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />



<title>Parallel Testing</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
</style>



<style type="text/css">
code {
white-space: pre;
}
.sourceCode {
overflow: visible;
}
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
{ counter-reset: source-line 0; }
pre.numberSource code > span
{ position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
{ content: counter(source-line);
position: relative; left: -1em; text-align: right; vertical-align: baseline;
border: none; display: inline-block;
-webkit-touch-callout: none; -webkit-user-select: none;
-khtml-user-select: none; -moz-user-select: none;
-ms-user-select: none; user-select: none;
padding: 0 4px; width: 4em;
color: #aaaaaa;
}
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa; padding-left: 4px; }
div.sourceCode
{ }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } 
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.at { color: #7d9029; } 
code span.bn { color: #40a070; } 
code span.bu { color: #008000; } 
code span.cf { color: #007020; font-weight: bold; } 
code span.ch { color: #4070a0; } 
code span.cn { color: #880000; } 
code span.co { color: #60a0b0; font-style: italic; } 
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.do { color: #ba2121; font-style: italic; } 
code span.dt { color: #902000; } 
code span.dv { color: #40a070; } 
code span.er { color: #ff0000; font-weight: bold; } 
code span.ex { } 
code span.fl { color: #40a070; } 
code span.fu { color: #06287e; } 
code span.im { color: #008000; font-weight: bold; } 
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.kw { color: #007020; font-weight: bold; } 
code span.op { color: #666666; } 
code span.ot { color: #007020; } 
code span.pp { color: #bc7a00; } 
code span.sc { color: #4070a0; } 
code span.ss { color: #bb6688; } 
code span.st { color: #4070a0; } 
code span.va { color: #19177c; } 
code span.vs { color: #4070a0; } 
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } 
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">Parallel Testing</h1>



<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="fu">library</span>(segtest)</span></code></pre></div>
<p>We provide a walk-through of how to run tests for segregation
distortion at many loci in parallel.</p>
<p>Data sets <code>ufit</code>, <code>ufit2</code>, and
<code>ufit3</code> contain the genotyping output of
<code>updog::multidog()</code> using three different models.</p>
<ul>
<li><code>ufit</code>: Uses the <code>norm</code> model without
specifying who the parents are.</li>
<li><code>ufit2</code>: Uses the <code>f1pp</code> model, specifying the
parents.</li>
<li><code>ufit3</code>: Uses the <code>f1</code> model, specifying the
parents.</li>
</ul>
<p>You can convert this genotyping output to what
<code>multi_lrt()</code> expects using <code>multidog_to_g()</code>.</p>
<p>If you did <em>not</em> use the <code>f1pp</code> or <code>f1</code>
models, use ether the <code>all_gl</code> (to run tests using genotype
log-likelihoods) or <code>all_g</code> (to run tests assuming genotypes
are known) options, and you must specify the ID’s of the parents.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a>o1 <span class="ot">&lt;-</span> <span class="fu">multidog_to_g</span>(ufit, <span class="at">type =</span> <span class="st">&quot;all_g&quot;</span>, <span class="at">p1 =</span> <span class="st">&quot;indigocrisp&quot;</span>, <span class="at">p2 =</span> <span class="st">&quot;sweetcrisp&quot;</span>)</span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a>o2 <span class="ot">&lt;-</span> <span class="fu">multidog_to_g</span>(ufit, <span class="at">type =</span> <span class="st">&quot;all_gl&quot;</span>, <span class="at">p1 =</span> <span class="st">&quot;indigocrisp&quot;</span>, <span class="at">p2 =</span> <span class="st">&quot;sweetcrisp&quot;</span>)</span></code></pre></div>
<p>If you <em>did</em> use the <code>f1pp</code> or <code>f1</code>
models, use either the <code>off_gl</code> (to run tests using genotype
log-likelihoods) or <code>off_g</code> (to run tests assuming genotypes
are known) options.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a>o3 <span class="ot">&lt;-</span> <span class="fu">multidog_to_g</span>(ufit2, <span class="at">type =</span> <span class="st">&quot;off_g&quot;</span>)</span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a>o4 <span class="ot">&lt;-</span> <span class="fu">multidog_to_g</span>(ufit2, <span class="at">type =</span> <span class="st">&quot;off_gl&quot;</span>)</span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a>o5 <span class="ot">&lt;-</span> <span class="fu">multidog_to_g</span>(ufit3, <span class="at">type =</span> <span class="st">&quot;off_g&quot;</span>)</span>
<span id="cb3-4"><a href="#cb3-4" tabindex="-1"></a>o6 <span class="ot">&lt;-</span> <span class="fu">multidog_to_g</span>(ufit3, <span class="at">type =</span> <span class="st">&quot;off_gl&quot;</span>)</span></code></pre></div>
<p>We would recommend <em>always</em> using genotype log-likelihoods.
But the option is there for known genotypes.</p>
<p>Parallelization support is provided through the future package. You
specify how to implement parallelization using
<code>future::plan()</code>. Then you run <code>multi_lrt()</code>. Then
you shut down the parallelization with <code>future::plan()</code>. The
most common plan is <code>future::multisession()</code>, where you
specify the number of parallel processes with the <code>workers</code>
argument. You can get the maximum number of workers via
<code>future::availableCores()</code></p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a>future<span class="sc">::</span><span class="fu">availableCores</span>()</span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a><span class="co">#&gt; system </span></span>
<span id="cb4-3"><a href="#cb4-3" tabindex="-1"></a><span class="co">#&gt;     16</span></span></code></pre></div>
<p>So a typically workload looks something like this:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a>future<span class="sc">::</span><span class="fu">plan</span>(future<span class="sc">::</span><span class="fu">multisession</span>(<span class="at">workers =</span> <span class="dv">2</span>))</span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a>mout <span class="ot">&lt;-</span> <span class="fu">multi_lrt</span>(<span class="at">g =</span> o2<span class="sc">$</span>g, <span class="at">p1 =</span> o2<span class="sc">$</span>p1, <span class="at">p2 =</span> o2<span class="sc">$</span>p2)</span>
<span id="cb5-3"><a href="#cb5-3" tabindex="-1"></a>future<span class="sc">::</span><span class="fu">plan</span>(future<span class="sc">::</span><span class="fu">sequential</span>())</span></code></pre></div>
<p>The output is a data frame. The most important parts of this are the
<code>snp</code> and <code>p_value</code> columns. As a reminder, please
ignore the <code>alpha</code>, <code>xi1</code>, and <code>xi2</code>
estimates. Those are noisy. Please don’t use them for real work.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a>mout[<span class="fu">c</span>(<span class="st">&quot;snp&quot;</span>, <span class="st">&quot;p_value&quot;</span>)]</span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a><span class="co">#&gt;            snp     p_value</span></span>
<span id="cb6-3"><a href="#cb6-3" tabindex="-1"></a><span class="co">#&gt; 1   1_44673475 0.205250847</span></span>
<span id="cb6-4"><a href="#cb6-4" tabindex="-1"></a><span class="co">#&gt; 2  11_28836119 0.200142085</span></span>
<span id="cb6-5"><a href="#cb6-5" tabindex="-1"></a><span class="co">#&gt; 3  12_31029646 0.018008460</span></span>
<span id="cb6-6"><a href="#cb6-6" tabindex="-1"></a><span class="co">#&gt; 4   12_8487773 1.000000000</span></span>
<span id="cb6-7"><a href="#cb6-7" tabindex="-1"></a><span class="co">#&gt; 5    2_1426329 0.453995964</span></span>
<span id="cb6-8"><a href="#cb6-8" tabindex="-1"></a><span class="co">#&gt; 6   2_20070837 0.001131421</span></span>
<span id="cb6-9"><a href="#cb6-9" tabindex="-1"></a><span class="co">#&gt; 7   2_40108108 0.410718596</span></span>
<span id="cb6-10"><a href="#cb6-10" tabindex="-1"></a><span class="co">#&gt; 8   4_37820371 0.212515165</span></span>
<span id="cb6-11"><a href="#cb6-11" tabindex="-1"></a><span class="co">#&gt; 9   6_30619745 0.617023681</span></span>
<span id="cb6-12"><a href="#cb6-12" tabindex="-1"></a><span class="co">#&gt; 10   6_4350249 0.649505305</span></span></code></pre></div>
<p>It looks like SNPs <em>12_31029646</em> and <em>2_20070837</em> are
in possible segregation distortion. Let’s look at the posterior mode
genotypes of SNP 2_20070837:</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a>o1<span class="sc">$</span>g[<span class="st">&quot;2_20070837&quot;</span>, ]</span>
<span id="cb7-2"><a href="#cb7-2" tabindex="-1"></a><span class="co">#&gt;   0   1   2   3   4 </span></span>
<span id="cb7-3"><a href="#cb7-3" tabindex="-1"></a><span class="co">#&gt;  40 166  32   2   0</span></span>
<span id="cb7-4"><a href="#cb7-4" tabindex="-1"></a>o1<span class="sc">$</span>p1[<span class="st">&quot;2_20070837&quot;</span>]</span>
<span id="cb7-5"><a href="#cb7-5" tabindex="-1"></a><span class="co">#&gt; 2_20070837 </span></span>
<span id="cb7-6"><a href="#cb7-6" tabindex="-1"></a><span class="co">#&gt;          0</span></span>
<span id="cb7-7"><a href="#cb7-7" tabindex="-1"></a>o1<span class="sc">$</span>p2[<span class="st">&quot;2_20070837&quot;</span>]</span>
<span id="cb7-8"><a href="#cb7-8" tabindex="-1"></a><span class="co">#&gt; 2_20070837 </span></span>
<span id="cb7-9"><a href="#cb7-9" tabindex="-1"></a><span class="co">#&gt;          2</span></span></code></pre></div>
<p>So SNP 2_20070837 likely got flagged because of two individuals that
are very likely a genotype of 3, which is impossible if the parents have
genotypes 0 and 2.</p>
<p>For SNP 12_31029646, let’s compare the expected frequencies against
the observed modes.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a><span class="fu">offspring_gf_2</span>(<span class="at">alpha =</span> <span class="fl">0.1249</span>, <span class="at">xi1 =</span> <span class="dv">1</span><span class="sc">/</span><span class="dv">3</span>, <span class="at">p1 =</span> <span class="dv">2</span>, <span class="at">p2 =</span> <span class="dv">0</span>)</span>
<span id="cb8-2"><a href="#cb8-2" tabindex="-1"></a><span class="co">#&gt; [1] 0.2083 0.5834 0.2083 0.0000 0.0000</span></span>
<span id="cb8-3"><a href="#cb8-3" tabindex="-1"></a>o1<span class="sc">$</span>g[<span class="st">&quot;12_31029646&quot;</span>, ] <span class="sc">/</span> <span class="fu">sum</span>(o1<span class="sc">$</span>g[<span class="st">&quot;12_31029646&quot;</span>, ])</span>
<span id="cb8-4"><a href="#cb8-4" tabindex="-1"></a><span class="co">#&gt;         0         1         2         3         4 </span></span>
<span id="cb8-5"><a href="#cb8-5" tabindex="-1"></a><span class="co">#&gt; 0.1541667 0.5833333 0.2625000 0.0000000 0.0000000</span></span>
<span id="cb8-6"><a href="#cb8-6" tabindex="-1"></a>o1<span class="sc">$</span>p1[<span class="st">&quot;12_31029646&quot;</span>]</span>
<span id="cb8-7"><a href="#cb8-7" tabindex="-1"></a><span class="co">#&gt; 12_31029646 </span></span>
<span id="cb8-8"><a href="#cb8-8" tabindex="-1"></a><span class="co">#&gt;           2</span></span>
<span id="cb8-9"><a href="#cb8-9" tabindex="-1"></a>o1<span class="sc">$</span>p2[<span class="st">&quot;12_31029646&quot;</span>]</span>
<span id="cb8-10"><a href="#cb8-10" tabindex="-1"></a><span class="co">#&gt; 12_31029646 </span></span>
<span id="cb8-11"><a href="#cb8-11" tabindex="-1"></a><span class="co">#&gt;           0</span></span></code></pre></div>
<p>So SNP 12_31029646 likely got flagged because there are too many
individuals with a genotype of 2, and not enough with a genotype of
0.</p>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
